source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,444 @@
+ * @param command_line_options The Nighthawk Service load proto, written if the step controller
+ * returned a value.
+ * @param output The adaptive load session output proto where any error code is stored.
+ *
+ * @return bool True if the step controller provided a value, false if an error occurred.
+ */
+bool GetCommandLineOptionsOrSetOutputError(
+    const StepController& step_controller,
+    nighthawk::client::CommandLineOptions& command_line_options,
+    AdaptiveLoadSessionOutput& output) {
+  absl::StatusOr<nighthawk::client::CommandLineOptions> command_line_options_or =
+      step_controller.GetCurrentCommandLineOptions();
+  if (!command_line_options_or.ok()) {
+    std::string message =
+        absl::StrCat("Error setting Nighthawk input: ", command_line_options_or.status().message());
+    output.mutable_session_status()->set_code(grpc::ABORTED);
+    output.mutable_session_status()->set_message(message);
+    ENVOY_LOG_MISC(info, message);
+    return false;
+  }
+  command_line_options = command_line_options_or.value();
+  return true;
+}
+
+} // namespace
+
+AdaptiveLoadSessionOutput PerformAdaptiveLoadSession(
+    nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
+    const AdaptiveLoadSessionSpec& input_spec, Envoy::TimeSource& time_source) {
+  AdaptiveLoadSessionOutput output;
+
+  AdaptiveLoadSessionSpec spec = SetDefaults(input_spec);
+  absl::Status validation_status = CheckSessionSpec(spec);
+  if (!validation_status.ok()) {
+    SetOutputError(static_cast<int>(validation_status.code()), validation_status.message(), output);
+    return output;
+  }
+
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map;
+  for (const envoy::config::core::v3::TypedExtensionConfig& config :
+       spec.metrics_plugin_configs()) {
+    absl::StatusOr<MetricsPluginPtr> metrics_plugin_or = LoadMetricsPlugin(config);
+    RELEASE_ASSERT(
+        metrics_plugin_or.ok(),
+        absl::StrCat(
+            "MetricsPlugin loading error should have been caught during input validation: ",
+            metrics_plugin_or.status().message()));
+    name_to_custom_metrics_plugin_map[config.name()] = std::move(metrics_plugin_or.value());
+  }
+
+  absl::StatusOr<StepControllerPtr> step_controller_or =
+      LoadStepControllerPlugin(spec.step_controller_config(), spec.nighthawk_traffic_template());
+  RELEASE_ASSERT(
+      step_controller_or.ok(),
+      absl::StrCat(
+          "StepController plugin loading error should have been caught during input validation: ",
+          step_controller_or.status().message()));
+  StepControllerPtr step_controller = std::move(step_controller_or.value());
+
+  for (const nighthawk::adaptive_load::MetricSpecWithThreshold& threshold :
+       spec.metric_thresholds()) {
+    *output.mutable_metric_thresholds()->Add() = threshold;
+  }
+
+  Envoy::MonotonicTime start_time = time_source.monotonicTime();
+  while (!step_controller->IsConverged()) {
```

Comment:
Assiuming we can't converge right away, I wonder if expressing this along the lines of:

```c++
do {
   ....
   terminated = step_controller->IsDoomed(doom_reason) || step_controller->IsConverged(); 
} while(!terminated);

if (!doom_reason.empty()) {
  // doom handling
}
```

would help clarity. entirely subjective -- please ignore if you feel otherwise.

Reply:
Done

Reply:
nice, thank you